The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9333495525/job/25690588291) [(plain)](https://github.com/rust-lang/rust/commit/c5273bf423c3386caf8e6aed547b36dd15e852e0/checks/25690588291/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:a5ac7e51b41094c92402da3b24376905380afc29)
Download action repository 'actions/setup-python@v5' (SHA:82c7e631bb3cdc910f68e0081d67478d79c6982d)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 CACHED

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 CACHED

---
DirectMap4k:      190400 kB
DirectMap2M:     6100992 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/0038c021031ce9f1ec2329469c8d85d0e681ef8f/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-0038c021031ce9f1ec2329469c8d85d0e681ef8f-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 24.93s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_mir_transform/src/validate.rs at line 1226:
                             "CastKind::{kind:?} output must be a raw pointer, not {:?}",
                         );
-                        }
+                    }
                     CastKind::PtrToPtr => {
                     CastKind::PtrToPtr => {
                         check_kinds!(
                             op_ty,
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/tests/codegen-units/item-collection/drop_in_place_intrinsic.rs" "/checkout/tests/codegen-units/item-collection/transitive-drop-glue.rs" "/checkout/tests/codegen-units/item-collection/tuple-drop-glue.rs" "/checkout/tests/codegen-units/polymorphization/poly-foreign.rs" "/checkout/tests/codegen-units/polymorphization/unused_type_parameters.rs" "/checkout/tests/codegen-units/polymorphization/auxiliary/poly-dep.rs" "/checkout/compiler/rustc_hir_pretty/src/lib.rs" "/checkout/compiler/rustc_mir_transform/src/coroutine/by_move_body.rs" "/checkout/compiler/rustc_mir_transform/src/large_enums.rs" "/checkout/compiler/rustc_mir_transform/src/ref_prop.rs" "/checkout/compiler/rustc_mir_transform/src/pass_manager.rs" "/checkout/compiler/rustc_mir_transform/src/dataflow_const_prop.rs" "/checkout/compiler/rustc_mir_transform/src/match_branches.rs" "/checkout/compiler/rustc_mir_transform/src/known_panics_lint.rs" "/checkout/compiler/rustc_mir_transform/src/add_subtyping_projections.rs" "/checkout/compiler/rustc_mir_transform/src/add_moves_for_packed_drops.rs" "/checkout/compiler/rustc_mir_transform/src/lib.rs" "/checkout/compiler/rustc_mir_transform/src/remove_noop_landing_pads.rs" "/checkout/compiler/rustc_mir_transform/src/inline.rs" "/checkout/compiler/rustc_mir_transform/src/early_otherwise_branch.rs" "/checkout/compiler/rustc_mir_transform/src/lint.rs" "/checkout/compiler/rustc_mir_transform/src/ssa.rs" "/checkout/compiler/rustc_mir_transform/src/jump_threading.rs" "/checkout/compiler/rustc_mir_transform/src/lower_slice_len.rs" "/checkout/compiler/rustc_mir_transform/src/sroa.rs" "/checkout/compiler/rustc_mir_transform/src/cross_crate_inline.rs" "/checkout/compiler/rustc_mir_transform/src/cost_checker.rs" "/checkout/compiler/rustc_mir_transform/src/validate.rs" "/checkout/compiler/rustc_mir_transform/src/add_call_guards.rs" "/checkout/compiler/rustc_mir_transform/src/dead_store_elimination.rs" "/checkout/compiler/rustc_mir_transform/src/ffi_unwind_calls.rs" "/checkout/compiler/rustc_mir_transform/src/check_const_item_mutation.rs" "/checkout/compiler/rustc_mir_transform/src/coroutine.rs" "/checkout/compiler/rustc_mir_transform/src/check_alignment.rs" "/checkout/compiler/rustc_mir_transform/src/cleanup_post_borrowck.rs" "/checkout/compiler/rustc_mir_transform/src/function_item_references.rs" "/checkout/compiler/rustc_mir_transform/src/dest_prop.rs" "/checkout/compiler/rustc_mir_transform/src/deduce_param_attrs.rs" "/checkout/compiler/rustc_mir_transform/src/elaborate_drops.rs" "/checkout/compiler/rustc_mir_transform/src/simplify_comparison_integral.rs" "/checkout/compiler/rustc_mir_transform/src/unreachable_prop.rs" "/checkout/compiler/rustc_mir_transform/src/shim.rs" "/checkout/compiler/rustc_mir_transform/src/prettify.rs" "/checkout/compiler/rustc_mir_transform/src/remove_zsts.rs" "/checkout/compiler/rustc_mir_transform/src/deref_separator.rs" "/checkout/compiler/rustc_mir_transform/src/shim/async_destructor_ctor.rs" "/checkout/compiler/rustc_mir_transform/src/ctfe_limit.rs" "/checkout/compiler/rustc_mir_transform/src/lower_intrinsics.rs" "/checkout/compiler/rustc_mir_transform/src/remove_storage_markers.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/query.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/mod.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/spans/from_mir.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/counters.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/tests.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/graph.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/mappings.rs" "/checkout/compiler/rustc_mir_transform/src/coverage/spans.rs" "/checkout/compiler/rustc_mir_transform/src/inline/cycle.rs" "/checkout/compiler/rustc_mir_transform/src/elaborate_box_derefs.rs" "/checkout/compiler/rustc_mir_transform/src/errors.rs" "/checkout/compiler/rustc_mir_transform/src/deduplicate_blocks.rs" "/checkout/compiler/rustc_mir_transform/src/remove_place_mention.rs" "/checkout/compiler/rustc_mir_transform/src/add_retag.rs" "/checkout/tests/codegen-units/item-collection/static-init.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Sat Jun  1 22:12:58 UTC 2024
  network time: Sat, 01 Jun 2024 22:12:58 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_mir_transform/src/validate.rs
+++ b/compiler/rustc_mir_transform/src/validate.rs
@@ -8,6 +8,7 @@ use rustc_middle::mir::coverage::CoverageKind;
 use rustc_middle::mir::interpret::Scalar;
 use rustc_middle::mir::visit::{NonUseContext, PlaceContext, Visitor};
 use rustc_middle::mir::*;
+use rustc_middle::ty::adjustment::PointerCoercion;
 use rustc_middle::ty::{
     self, CoroutineArgsExt, InstanceDef, ParamEnv, Ty, TyCtxt, TypeVisitableExt, Variance,
 };
@@ -1134,9 +1135,76 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                         // FIXME(dyn-star): make sure nothing needs to be done here.
                     }
                     // FIXME: Add Checks for these
-                    CastKind::PointerWithExposedProvenance
-                    | CastKind::PointerExposeProvenance
-                    | CastKind::PointerCoercion(_) => {}
+                    CastKind::PointerWithExposedProvenance | CastKind::PointerExposeProvenance => {}
+                    CastKind::PointerCoercion(PointerCoercion::ReifyFnPointer) => {
+                        // FIXME: check signature compatibility.
+                        check_kinds!(
+                            op_ty,
+                            "CastKind::{kind:?} input must be a fn item, not {:?}",
+                            ty::FnDef(..)
+                        );
+                        check_kinds!(
+                            target_type,
+                            "CastKind::{kind:?} output must be a fn pointer, not {:?}",
+                            ty::FnPtr(..)
+                        );
+                    }
+                    CastKind::PointerCoercion(PointerCoercion::UnsafeFnPointer) => {
+                        // FIXME: check safety and signature compatibility.
+                        check_kinds!(
+                            op_ty,
+                            "CastKind::{kind:?} input must be a fn pointer, not {:?}",
+                            ty::FnPtr(..)
+                        );
+                        check_kinds!(
+                            target_type,
+                            "CastKind::{kind:?} output must be a fn pointer, not {:?}",
+                            ty::FnPtr(..)
+                        );
+                    }
+                    CastKind::PointerCoercion(PointerCoercion::ClosureFnPointer(..)) => {
+                        // FIXME: check safety, captures, and signature compatibility.
+                        check_kinds!(
+                            op_ty,
+                            "CastKind::{kind:?} input must be a closure, not {:?}",
+                            ty::Closure(..)
+                        );
+                        check_kinds!(
+                            target_type,
+                            "CastKind::{kind:?} output must be a fn pointer, not {:?}",
+                            ty::FnPtr(..)
+                        );
+                    }
+                    CastKind::PointerCoercion(PointerCoercion::MutToConstPointer) => {
+                        // FIXME: check same pointee?
+                        check_kinds!(
+                            op_ty,
+                            "CastKind::{kind:?} input must be a raw mut pointer, not {:?}",
+                            ty::RawPtr(_, Mutability::Mut)
+                        );
+                        check_kinds!(
+                            target_type,
+                            "CastKind::{kind:?} output must be a raw const pointer, not {:?}",
+                            ty::RawPtr(_, Mutability::Not)
+                        );
+                    }
+                    CastKind::PointerCoercion(PointerCoercion::ArrayToPointer) => {
+                        // FIXME: Check pointee types
+                        check_kinds!(
+                            op_ty,
+                            "CastKind::{kind:?} input must be a raw pointer, not {:?}",
+                            ty::RawPtr(..)
+                        );
+                        check_kinds!(
+                            target_type,
+                            "CastKind::{kind:?} output must be a raw pointer, not {:?}",
+                            ty::RawPtr(..)
+                        );
+                    }
+                    CastKind::PointerCoercion(PointerCoercion::Unsize) => {
+                        // This is used for all `CoerceUnsized` types,
+                        // not just pointers/references, so is hard to check.
+                    }
                     CastKind::IntToInt | CastKind::IntToFloat => {
                         let input_valid = op_ty.is_integral() || op_ty.is_char() || op_ty.is_bool();
                         let target_valid = target_type.is_numeric() || target_type.is_char();
@@ -1147,10 +1215,29 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                             );
                         }
                     }
-                    CastKind::FnPtrToPtr | CastKind::PtrToPtr => {
-                        if !(op_ty.is_any_ptr() && target_type.is_unsafe_ptr()) {
-                            self.fail(location, "Can't cast {op_ty} into 'Ptr'");
-                        }
+                    CastKind::FnPtrToPtr => {
+                        check_kinds!(
+                            op_ty,
+                            "CastKind::{kind:?} input must be a fn pointer, not {:?}",
+                            ty::FnPtr(..)
+                        );
+                        check_kinds!(
+                            target_type,
+                            "CastKind::{kind:?} output must be a raw pointer, not {:?}",
+                            ty::RawPtr(..)
+                        );
+                    }
+                    CastKind::PtrToPtr => {
+                        check_kinds!(
+                            op_ty,
+                            "CastKind::{kind:?} input must be a raw pointer, not {:?}",
+                            ty::RawPtr(..)
+                        );
+                        check_kinds!(
+                            target_type,
+                            "CastKind::{kind:?} output must be a raw pointer, not {:?}",
+                            ty::RawPtr(..)
+                        );
                     }
                     CastKind::FloatToFloat | CastKind::FloatToInt => {
                         if !op_ty.is_floating_point() || !target_type.is_numeric() {

